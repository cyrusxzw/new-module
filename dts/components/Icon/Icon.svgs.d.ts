import type { Svg } from './Icon.types';
declare const svgs: Svg[];
export { svgs };
