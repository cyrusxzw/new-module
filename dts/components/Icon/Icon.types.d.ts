import type { Themes } from '~/types';
declare type SvgName =
  | ''
  | 'rightArrow'
  | 'warning'
  | 'rightUpArrow'
  | 'downArrow'
  | 'plusAndCloseWithCircle'
  | 'chevron'
  | 'close'
  | 'play'
  | 'pause'
  | 'muted'
  | 'unmuted'
  | 'aesop'
  | 'download'
  | 'seek'
  | 'scrolldown';
declare type Svg = {
  data: {
    [key: string]: unknown;
  }[];
  height?: string;
  name: SvgName;
  viewBox: string;
  width?: string;
};
declare type IconProps = {
  className?: string;
  dataRef?: string;
  height?: number;
  isActive?: boolean;
  name: SvgName;
  tabIndex?: number;
  theme?: Themes;
  title?: string;
  width?: number;
};
export type { IconProps, SvgName, Svg };
