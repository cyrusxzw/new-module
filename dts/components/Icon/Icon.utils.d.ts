import { ReactElement } from 'react';
import type { Svg } from './Icon.types';
declare const generateSvgBlueprint: (
  svg: Svg,
  uidKey: string,
) => ReactElement[];
export { generateSvgBlueprint };
