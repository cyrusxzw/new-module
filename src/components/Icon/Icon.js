import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { useUID } from 'react-uid';
import find from 'lodash/find';
import { svgs } from './Icon.svgs';
import { generateSvgBlueprint } from './Icon.utils';
import styles from './Icon.module.css';

const Icon = ({
  className,
  dataRef,
  height = 12,
  isActive = false,
  name,
  tabIndex = -1,
  theme = 'dark',
  title,
  width = 12,
}) => {
  const uid = useUID();
  const svg = find(svgs, { name });

  if (svg === undefined) {
    return null;
  }

  const ariaTitle = title || name;
  const uidAriaLabellBy = `${name}-${uid}`;
  const svgBlueprint = generateSvgBlueprint(svg, uid);
  const classSet = cx(
    styles.base,
    styles[name],
    styles[theme],
    {
      [styles.isActive]: isActive,
    },
    className,
  );

  return (
    <svg
      aria-labelledby={uidAriaLabellBy}
      className={classSet}
      data-ref={dataRef}
      data-testid="data-testid-Icon"
      focusable="false"
      height={height}
      role="img"
      style={{ height: `${height}px`, width: `${width}px` }}
      tabIndex={tabIndex}
      viewBox={svg.viewBox}
      width={width}
    >
      <title id={uidAriaLabellBy}>{ariaTitle}</title>
      <g>{svgBlueprint}</g>
    </svg>
  );
};

Icon.propTypes = {
  className: PropTypes.string,
  dataRef: PropTypes.string,
  height: PropTypes.number,
  isActive: PropTypes.bool,
  name: PropTypes.string.isRequired,
  tabIndex: PropTypes.number,
  theme: PropTypes.oneOf(['dark', 'light']),
  title: PropTypes.string,
  width: PropTypes.number,
};

export { Icon };
