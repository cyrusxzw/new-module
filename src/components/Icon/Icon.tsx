import React, { FC } from 'react';
import cx from 'classnames';
import { useUID } from 'react-uid';
import find from 'lodash/find';
import { svgs } from './Icon.svgs';
import { generateSvgBlueprint } from './Icon.utils';
import type { IconProps } from './Icon.types';
import styles from './Icon.module.css';

const Icon: FC<IconProps> = ({
  className,
  dataRef,
  height = 12,
  isActive = false,
  name,
  tabIndex = -1,
  theme = 'dark',
  title,
  width = 12,
}) => {
  const uid = useUID();
  const svg = find(svgs, { name });

  if (svg === undefined) {
    return null;
  }

  const ariaTitle = title || name;
  const uidAriaLabelBy = `${name}-${uid}`;
  const svgBlueprint = generateSvgBlueprint(svg, uid);
  const classSet = cx(
    styles.base,
    styles[name],
    styles[theme],
    {
      [styles.isActive]: isActive,
    },
    className,
  );

  return (
    <svg
      aria-labelledby={uidAriaLabelBy}
      className={classSet}
      data-ref={dataRef}
      data-testid="data-testid-Icon"
      focusable="false"
      height={height}
      role="img"
      style={{ height: `${height}px`, width: `${width}px` }}
      tabIndex={tabIndex}
      viewBox={svg.viewBox}
      width={width}
    >
      <title id={uidAriaLabelBy}>{ariaTitle}</title>
      <g>{svgBlueprint}</g>
    </svg>
  );
};

export { Icon };
