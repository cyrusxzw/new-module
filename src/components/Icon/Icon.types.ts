import type { Themes } from '~/types';

type SvgName =
  | ''
  | 'rightArrow'
  | 'warning'
  | 'rightUpArrow'
  | 'downArrow'
  | 'plusAndCloseWithCircle'
  | 'chevron'
  | 'close'
  | 'play'
  | 'pause'
  | 'muted'
  | 'unmuted'
  | 'aesop'
  | 'download'
  | 'seek'
  | 'scrolldown';

type Svg = {
  data: { [key: string]: unknown }[];
  height?: string;
  name: SvgName;
  viewBox: string;
  width?: string;
};

type IconProps = {
  className?: string;
  dataRef?: string;
  height?: number;
  isActive?: boolean;
  name: SvgName;
  tabIndex?: number;
  theme?: Themes;
  title?: string;
  width?: number;
};

export type { IconProps, SvgName, Svg };
