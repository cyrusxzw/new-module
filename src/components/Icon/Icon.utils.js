import React from 'react';
import styles from './Icon.module.css';

const generateSvgBlueprint = (svg, uidKey) => {
  function mapSvgData(svgData) {
    return svgData.map((item, i) => {
      const tagName = Object.keys(item)[0];
      const { classes, data, ...props } = item[tagName];

      props.key = `${uidKey}${i}`;

      if (classes !== undefined) {
        props.className = styles[classes];
      }

      if (tagName === 'g') {
        return React.createElement(tagName, props, mapSvgData(data));
      }

      return React.createElement(tagName, props);
    });
  }

  return mapSvgData(svg.data);
};

export { generateSvgBlueprint };
