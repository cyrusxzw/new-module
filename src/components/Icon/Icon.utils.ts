import React, { ReactElement } from 'react';
import type { Svg } from './Icon.types';
import styles from './Icon.module.css';

const generateSvgBlueprint = (svg: Svg, uidKey: string): ReactElement[] => {
  function mapSvgData(svgData): ReactElement[] {
    return svgData.map((item, i) => {
      const tagName = Object.keys(item)[0];
      const { classes, data, ...props } = item[tagName];

      props.key = `${uidKey}${i}`;

      if (classes !== undefined) {
        props.className = styles[classes];
      }

      if (tagName === 'g') {
        return React.createElement(tagName, props, mapSvgData(data));
      }

      return React.createElement(tagName, props);
    });
  }

  return mapSvgData(svg.data);
};

export { generateSvgBlueprint };
