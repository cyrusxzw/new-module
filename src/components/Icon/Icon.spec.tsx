import React from 'react';
import { render, screen } from '@testing-library/react';
import { Icon } from './Icon';

describe('<Icon />', () => {
  it('should render render base component correctly', () => {
    const { container } = render(<Icon name="rightArrow" />);

    expect(container).toMatchSnapshot();
  });

  it('should render SVG with provided title prop correctly', () => {
    render(<Icon name="chevron" title="title" />);

    const icon = screen.getByRole('img', { name: /title/i });

    expect(icon).toBeInTheDocument();
  });

  it('should render SVG that has nested `g` tags correctly', () => {
    const { container } = render(<Icon name="muted" title="title" />);

    expect(container).toMatchSnapshot();
  });

  it('should return `null` if svg name is not found', () => {
    render(<Icon name="" />);

    const child = screen.queryByTestId(/data-testid-Icon/i);

    expect(child).not.toBeInTheDocument();
  });
});
