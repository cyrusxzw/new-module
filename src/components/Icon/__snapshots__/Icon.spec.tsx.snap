// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> should render SVG that has nested \`g\` tags correctly 1`] = `
<div>
  <svg
    aria-labelledby="muted-3"
    class="base muted dark"
    data-testid="data-testid-Icon"
    focusable="false"
    height="12"
    role="img"
    style="height: 12px; width: 12px;"
    tabindex="-1"
    viewBox="8 7 12 18"
    width="12"
  >
    <title
      id="muted-3"
    >
      title
    </title>
    <g>
      <g
        transform="translate(-97.000000, -48.000000)"
      >
        <g
          transform="translate(97.000000, 48.000000)"
        >
          <g
            fill="#FFFEF2"
            transform="translate(4.000000, 6.000000)"
          >
            <g
              transform="translate(3.000000, 0.573593)"
            >
              <path
                d="M5.74011671,4.90798034 L13.5001929,12.6680565 C13.8907172,13.0585808 13.8907172,13.6917458 13.5001929,14.0822701 C13.3186245,14.2638385 13.0741693,14.368605 12.8174691,14.374866 L3.81370319,14.59447 L4.03330724,5.59070413 C4.0467736,5.03858358 4.50527237,4.60191807 5.05739292,4.61538443 C5.31409312,4.62164541 5.55854828,4.72641191 5.74011671,4.90798034 Z"
                transform="translate(9.481906, 8.926267) rotate(45.000000) translate(-9.481906, -8.926267)"
              />
              <polygon
                points="0 7.4516129 4.4516129 6.16129032 4.4516129 11.6912442 0 10.4009217"
              />
            </g>
            <path
              d="M18.353198,12.259593 C18.3633576,11.843048 18.3297871,11.4377194 18.2566733,11.0477941 C17.7619986,8.4096316 15.457178,6.47656906 12.6389123,6.54530725 C9.4041016,6.62420507 18.2743002,15.4944036 18.353198,12.259593 Z"
              transform="translate(15.142166, 9.756339) rotate(45.000000) translate(-15.142166, -9.756339)"
            />
            <rect
              height="25.5"
              rx="1"
              stroke="#858480"
              stroke-width="0.5"
              transform="translate(11.575976, 8.735040) scale(-1, 1) rotate(58.000000) translate(-11.575976, -8.735040)"
              width="2.5"
              x="10.325976"
              y="-4.01495963"
            />
          </g>
        </g>
      </g>
    </g>
  </svg>
</div>
`;

exports[`<Icon /> should render render base component correctly 1`] = `
<div>
  <svg
    aria-labelledby="rightArrow-1"
    class="base rightArrow dark"
    data-testid="data-testid-Icon"
    focusable="false"
    height="12"
    role="img"
    style="height: 12px; width: 12px;"
    tabindex="-1"
    viewBox="0 0 50 50"
    width="12"
  >
    <title
      id="rightArrow-1"
    >
      rightArrow
    </title>
    <g>
      <path
        d="M30.1,5.3L50,25.1L30.1,45h-6.6l18-17.6H0v-4.8h41.5l-18-17.6h6.6V5.3z"
      />
    </g>
  </svg>
</div>
`;
